apollo-normalized-cache-sqlite/src/iosMain/kotlin/com/apollographql/apollo/cache/normalized/sql/internal/DatabaseRequestExecutor.kt

```diff
@@ -0,0 +1,83 @@
+package com.apollographql.apollo.cache.normalized.sql.internal
+
+import com.apollographql.apollo.cache.normalized.sql.ApolloDatabase
+import kotlinx.cinterop.COpaquePointer
+import kotlinx.cinterop.StableRef
+import kotlinx.cinterop.asStableRef
+import kotlinx.cinterop.autoreleasepool
+import kotlinx.cinterop.staticCFunction
+import platform.Foundation.NSThread
+import platform.darwin.dispatch_async_f
+import platform.darwin.dispatch_get_main_queue
+import kotlin.coroutines.Continuation
+import kotlin.coroutines.suspendCoroutine
+import kotlin.native.concurrent.TransferMode
+import kotlin.native.concurrent.Worker
+import kotlin.native.concurrent.freeze
+
+/**
+ * Performs database operation on background thread via [Worker] API.
+ * This executor requires caller thread to be main.
+ * After database operation execution it resumes on main thread.
+ */
+internal actual class DatabaseRequestExecutor actual constructor(
+    private val database: ApolloDatabase
+) {
+  private val worker: Worker = Worker.start()
+
+  actual suspend fun <R> execute(operation: ApolloDatabase.() -> R): R {
+    return suspendCoroutine { continuation ->
+      Request(
+          worker = worker,
+          database = database,
+          operation = operation,
+          continuation = continuation,
+      ).execute()
+    }
+  }
+
+  private class Request<R>(
+      val worker: Worker,
+      val database: ApolloDatabase,
+      val operation: ApolloDatabase.() -> R,
+      val continuation: Continuation<R>
+  ) {
+    fun execute() {
+      assert(NSThread.isMainThread())
+      val continuationPtr = StableRef.create(continuation).asCPointer()
+      worker.execute(
+          mode = TransferMode.SAFE,
+          producer = { (continuationPtr to { operation(database) }).freeze() },
```

Comment:
If I understand correctly, this will freeze the `ApolloDatabase` ? If that's the case, I'm amazed SQLDelight makes everything immutable 👍 . That subgraph must be quite a beast.

Reply:
Yeah, I think it's ok to freeze DB

Reply:
❄️ ⛄ ❄️ ⛄ ❄️ ⛄